In [1]:
# importing pandas as pd
import pandas as pd

In [2]:
# reading ExcelFile.
xls = pd.ExcelFile('Data.xlsx')

# preview of sheet names.
xls.sheet_names

['RH K-4 MOY Data',
 'MOY Export',
 'Syracuse',
 'Howard',
 'Swarthmore',
 'Spelman',
 'Williams',
 'Brooklyn',
 'Tuskegee',
 'Northwestern',
 'Georgetown',
 'Wisconsin']

In [3]:
df = pd.concat(pd.read_excel(xls, sheet_name=xls.sheet_names[2:]), ignore_index=True)
print(df.shape)
df.head()

(274, 17)


Student_ Number   Grade_Level  Home_room  BOY Indep Level  \
0       236563003.0           0.0   Syracuse              AA   
1       238194344.0           0.0   Syracuse              AA   
2       239980246.0           0.0   Syracuse              AA   
3       240845834.0           0.0   Syracuse              AA   
4       239980196.0           0.0   Syracuse              AA   

    Grade Level Equiv  Inst Level  Grade Level Equiv.1 MOY Indep Level  \
0               -0.25         NaN                                   AA   
1               -0.25         NaN                                   AA   
2               -0.25         NaN                                   AA   
3               -0.25         NaN                                   AA   
4               -0.25         NaN                                   AA   

   Grade Level Equiv.2  Inst Level.1  Grade Level Equiv.3   Accuracy (%)  \
0                -0.25             A                  0.0           88.0   
1                -0.25             A                  0.0           97.0   
2                -0.25             A                  0.0           33.0   
3                                 AA                -0.25            NaN   
4                                 AA                -0.25            NaN   

    Fluency (0-3)   Comp Within (0-3)   Comp Beyond (0-3)  Comp About  \
0             1.0                 3.0                 2.0         NaN   
1             2.0                 2.0                 1.0         NaN   
2             2.0                 1.0                 2.0         NaN   
3             3.0                 1.0                 NaN         NaN   
4             2.0                 3.0                 1.0         NaN   

    Total Comp  
0          5.0  
1          3.0  
2          3.0  
3          1.0  
4          4.0

In [4]:
# previewing df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Student_ Number       274 non-null    float64
 1   Grade_Level           274 non-null    float64
 2   Home_room             274 non-null    object 
 3   BOY Indep Level       273 non-null    object 
 4    Grade Level Equiv    259 non-null    float64
 5    Inst Level           187 non-null    object 
 6    Grade Level Equiv.1  274 non-null    object 
 7   MOY Indep Level       271 non-null    object 
 8    Grade Level Equiv.2  274 non-null    object 
 9    Inst Level.1         264 non-null    object 
 10   Grade Level Equiv.3  254 non-null    object 
 11   Accuracy (%)         218 non-null    float64
 12   Fluency (0-3)        232 non-null    float64
 13   Comp Within (0-3)    251 non-null    float64
 14   Comp Beyond (0-3)    249 non-null    float64
 15   Comp About           1

A couple of noticings:

* The column names are inconsistently formatted and need to be cleaned.
* Some numerical columns need to be casted as such.
* Since the Non-null values are not consistent, further investigation is needed.

In [5]:
# removing spaces and special characters from column names 
df.columns = df.columns.str.strip()
df.columns = df.columns.str.replace('_',' ')

df.insert(1,'Student Number', df['Student  Number'])
df.insert(2,'Homeroom', df['Home room'])

# columns to drop
drop_cols = ['Home room','Student  Number']
# drop columns
df.drop(drop_cols, axis=1, inplace=True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Student Number       274 non-null    float64
 1   Homeroom             274 non-null    object 
 2   Grade Level          274 non-null    float64
 3   BOY Indep Level      273 non-null    object 
 4   Grade Level Equiv    259 non-null    float64
 5   Inst Level           187 non-null    object 
 6   Grade Level Equiv.1  274 non-null    object 
 7   MOY Indep Level      271 non-null    object 
 8   Grade Level Equiv.2  274 non-null    object 
 9   Inst Level.1         264 non-null    object 
 10  Grade Level Equiv.3  254 non-null    object 
 11  Accuracy (%)         218 non-null    float64
 12  Fluency (0-3)        232 non-null    float64
 13  Comp Within (0-3)    251 non-null    float64
 14  Comp Beyond (0-3)    249 non-null    float64
 15  Comp About           132 non-null    obj

In [7]:
# get summary of missing values
df.isna().sum()

Student Number           0
Homeroom                 0
Grade Level              0
BOY Indep Level          1
Grade Level Equiv       15
Inst Level              87
Grade Level Equiv.1      0
MOY Indep Level          3
Grade Level Equiv.2      0
Inst Level.1            10
Grade Level Equiv.3     20
Accuracy (%)            56
Fluency (0-3)           42
Comp Within (0-3)       23
Comp Beyond (0-3)       25
Comp About             142
Total Comp               0
dtype: int64

For BOY and MOY missing data, I will use the key provided with this analysis.

In [8]:
# importing BOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1026408630'
url1 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_levels = pd.read_csv(url1, header=1)
df_levels.head(27)

Independent Reading Level  Grade Equivalent                  K  \
0                         AA             -0.25     On Grade Level   
1                          A              0.00  Above Grade Level   
2                          B              0.33  Above Grade Level   
3                          C              0.67  Above Grade Level   
4                          D              1.00  Above Grade Level   
5                          E              1.16  Above Grade Level   
6                          F              1.33  Above Grade Level   
7                          G              1.50  Above Grade Level   
8                          H              1.67  Above Grade Level   
9                          I              1.83  Above Grade Level   
10                         J              2.00  Above Grade Level   
11                         K              2.33  Above Grade Level   
12                         L              2.67  Above Grade Level   
13                         M              3.00  Above Grade Level   
14                         N              3.33  Above Grade Level   
15                         O              3.67  Above Grade Level   
16                         P              4.00  Above Grade Level   
17                         Q              4.33  Above Grade Level   
18                         R              4.67  Above Grade Level   
19                         S              5.00  Above Grade Level   
20                         T              5.33  Above Grade Level   
21                         U              5.67  Above Grade Level   
22                         V              6.00  Above Grade Level   
23                         W              6.33  Above Grade Level   
24                         X              6.67  Above Grade Level   
25                         Y              7.00  Above Grade Level   
26                         Z              8.00  Above Grade Level   

                          1                        2                        3  \
0         Below Grade Level        Below Grade Level        Below Grade Level   
1         Below Grade Level        Below Grade Level        Below Grade Level   
2         Below Grade Level        Below Grade Level        Below Grade Level   
3   Approaching Grade Level        Below Grade Level        Below Grade Level   
4            On Grade Level        Below Grade Level        Below Grade Level   
5         Above Grade Level        Below Grade Level        Below Grade Level   
6         Above Grade Level        Below Grade Level        Below Grade Level   
7         Above Grade Level        Below Grade Level        Below Grade Level   
8         Above Grade Level        Below Grade Level        Below Grade Level   
9         Above Grade Level  Approaching Grade Level        Below Grade Level   
10        Above Grade Level           On Grade Level        Below Grade Level   
11        Above Grade Level        Above Grade Level        Below Grade Level   
12        Above Grade Level        Above Grade Level  Approaching Grade Level   
13        Above Grade Level        Above Grade Level           On Grade Level   
14        Above Grade Level        Above Grade Level        Above Grade Level   
15        Above Grade Level        Above Grade Level        Above Grade Level   
16        Above Grade Level        Above Grade Level        Above Grade Level   
17        Above Grade Level        Above Grade Level        Above Grade Level   
18        Above Grade Level        Above Grade Level        Above Grade Level   
19        Above Grade Level        Above Grade Level        Above Grade Level   
20        Above Grade Level        Above Grade Level        Above Grade Level   
21        Above Grade Level        Above Grade Level        Above Grade Level   
22        Above Grade Level        Above Grade Level        Above Grade Level   
23        Above Grade Level        Above Grade Level        Above Grade Level   
24        Above Grade Level        Above Grad

In [9]:
# Importing MOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1314641425'
url2 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_MOY = pd.read_csv(url2, header=1)
df_MOY.head(27)

Independent Reading Level  Grade Equivalent                        K  \
0                         AA             -0.25        Below Grade Level   
1                          A              0.00  Approaching Grade Level   
2                          B              0.33           On Grade Level   
3                          C              0.67        Above Grade Level   
4                          D              1.00        Above Grade Level   
5                          E              1.16        Above Grade Level   
6                          F              1.33        Above Grade Level   
7                          G              1.50        Above Grade Level   
8                          H              1.67        Above Grade Level   
9                          I              1.83        Above Grade Level   
10                         J              2.00        Above Grade Level   
11                         K              2.33        Above Grade Level   
12                         L              2.67        Above Grade Level   
13                         M              3.00        Above Grade Level   
14                         N              3.33        Above Grade Level   
15                         O              3.67        Above Grade Level   
16                         P              4.00        Above Grade Level   
17                         Q              4.33        Above Grade Level   
18                         R              4.67        Above Grade Level   
19                         S              5.00        Above Grade Level   
20                         T              5.33        Above Grade Level   
21                         U              5.67        Above Grade Level   
22                         V              6.00        Above Grade Level   
23                         W              6.33        Above Grade Level   
24                         X              6.67        Above Grade Level   
25                         Y              7.00        Above Grade Level   
26                         Z              8.00        Above Grade Level   

                          1                        2                        3  \
0         Below Grade Level        Below Grade Level        Below Grade Level   
1         Below Grade Level        Below Grade Level        Below Grade Level   
2         Below Grade Level        Below Grade Level        Below Grade Level   
3         Below Grade Level        Below Grade Level        Below Grade Level   
4         Below Grade Level        Below Grade Level        Below Grade Level   
5         Below Grade Level        Below Grade Level        Below Grade Level   
6   Approaching Grade Level        Below Grade Level        Below Grade Level   
7            On Grade Level        Below Grade Level        Below Grade Level   
8         Above Grade Level        Below Grade Level        Below Grade Level   
9         Above Grade Level        Below Grade Level        Below Grade Level   
10        Above Grade Level  Approaching Grade Level        Below Grade Level   
11        Above Grade Level           On Grade Level        Below Grade Level   
12        Above Grade Level        Above Grade Level        Below Grade Level   
13        Above Grade Level        Above Grade Level  Approaching Grade Level   
14        Above Grade Level        Above Grade Level           On Grade Level   
15        Above Grade Level        Above Grade Level        Above Grade Level   
16        Above Grade Level        Above Grade Level        Above Grade Level   
17        Above Grade Level        Above Grade Level        Above Grade Level   
18        Above Grade Level        Above Grade Level        Above Grade Level   
19        Above Grade Level        Above Grade Level        Above Grade Level   
20        Above Grade Level        Above Grade Level        Above Grade Level   
21        Above Grade Level        Above Grade Level        Above Grade Level   
22        Above Grade Level        Abov

BOY and MOY Independent Reading Levels & Grade Equivalents are identical.  I will create a dictionary with these level to assist with missing values.

In [10]:
# Dictionary 
alpha_keys = list(df_levels['Independent Reading Level'].unique())
alpha_values = list(df_levels['Grade Equivalent'].unique())

levels = dict(zip(alpha_keys, alpha_values))
levels

{'AA': -0.25,
 'A': 0.0,
 'B': 0.33,
 'C': 0.67,
 'D': 1.0,
 'E': 1.16,
 'F': 1.33,
 'G': 1.5,
 'H': 1.67,
 'I': 1.83,
 'J': 2.0,
 'K': 2.33,
 'L': 2.67,
 'M': 3.0,
 'N': 3.33,
 'O': 3.67,
 'P': 4.0,
 'Q': 4.33,
 'R': 4.67,
 'S': 5.0,
 'T': 5.33,
 'U': 5.67,
 'V': 6.0,
 'W': 6.33,
 'X': 6.67,
 'Y': 7.0,
 'Z': 8.0}

In [11]:
# locating records with missing BOY Grade Equivalent data
print(df.loc[df['Grade Level Equiv'].isna()].shape)
df.loc[df['Grade Level Equiv'].isna()]

(15, 17)


Student Number      Homeroom  Grade Level BOY Indep Level  \
24      240415372.0      Syracuse          0.0                   
26      239668965.0      Syracuse          0.0                   
53      240905430.0        Howard          0.0                   
60      205888340.0    Swarthmore          1.0                   
67      238231914.0    Swarthmore          1.0             NaN   
73      237499462.0    Swarthmore          1.0                   
82      234292928.0    Swarthmore          1.0                   
83      218702165.0    Swarthmore          1.0                   
138     234743243.0      Williams          2.0                   
139     234411106.0      Williams          2.0                   
175     226787604.0      Tuskegee          3.0                   
189     233199898.0      Tuskegee          3.0                   
190     231197146.0      Tuskegee          3.0                   
195     203306691.0  Northwestern          3.0               H   
230     232795195.0    Georgetown          4.0                   

     Grade Level Equiv Inst Level Grade Level Equiv.1 MOY Indep Level  \
24                 NaN                                             AA   
26                 NaN                                             AA   
53                 NaN                                                  
60                 NaN                                              C   
67                 NaN        NaN                                   D   
73                 NaN                                              F   
82                 NaN                                                  
83                 NaN                                                  
138                NaN                                                  
139                NaN                                                  
175                NaN                                              M   
189                NaN                                                  
190                NaN                                                  
195                NaN                                              I   
230                NaN                                              Q   

    Grade Level Equiv.2 Inst Level.1 Grade Level Equiv.3  Accuracy (%)  \
24                                 A                 NaN           NaN   
26                                 A                 NaN           NaN   
53                                                   NaN           NaN   
60                 0.67            D                 1.0          99.0   
67                  1.0            E                1.16           NaN   
73                 1.33            G                 1.5          94.0   
82                                                   NaN           NaN   
83                                                   NaN           NaN   
138                                                  NaN           NaN   
139                                                  NaN           NaN   
175                 3.0            N                 NaN          99.0   
189                                                  NaN           NaN   
190                                                  NaN           NaN   
195                                J                 NaN          98.0   
230                4.33            R                4.67          99.0   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
24             NaN                NaN                NaN              
26             NaN                NaN                NaN              
53             NaN                NaN                NaN              
60             2.0                2.0                2.0        NaN   
67             1.0                1.0                2.0        NaN   
73             1.0                NaN                1.0        NaN   
82             NaN                NaN                NaN              
83             NaN        

Upon further investigation, only one BOY Independent Level was included (level H). My assumption is that these students do not have beginning year data because they enrolled in school after school had started. I will address the missing equivalency for the one student with data next. Afterwards, I will fill those values with NE for "Not Enrolled." 

In [12]:
# replacing grade equivalence for level H
df.loc[(df['BOY Indep Level']=='H'),'Grade Level Equiv']=1.67

# checking to see if change was made.
df.loc[df['Grade Level Equiv'].isna()].shape

(14, 17)

In [13]:
# Locating records with missing MOY Grade Equivalent data
print(df.loc[df['Grade Level Equiv.2'].isna()].shape)
df.loc[df['Grade Level Equiv.2'].isna()]

(0, 17)


Empty DataFrame
Columns: [Student Number, Homeroom, Grade Level, BOY Indep Level, Grade Level Equiv, Inst Level, Grade Level Equiv.1, MOY Indep Level, Grade Level Equiv.2, Inst Level.1, Grade Level Equiv.3, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [14]:
# Locating records with missing BOY Inst Level data
df.loc[df['Inst Level'].isna()]

Student Number      Homeroom  Grade Level BOY Indep Level  \
0       236563003.0      Syracuse          0.0              AA   
1       238194344.0      Syracuse          0.0              AA   
2       239980246.0      Syracuse          0.0              AA   
3       240845834.0      Syracuse          0.0              AA   
4       239980196.0      Syracuse          0.0              AA   
..              ...           ...          ...             ...   
213     237900477.0  Northwestern          3.0               P   
232     223455205.0    Georgetown          4.0               Q   
234     223474693.0    Georgetown          4.0               P   
239     205734916.0    Georgetown          4.0               R   
244     233255389.0    Georgetown          4.0               U   

     Grade Level Equiv Inst Level Grade Level Equiv.1 MOY Indep Level  \
0                -0.25        NaN                                  AA   
1                -0.25        NaN                                  AA   
2                -0.25        NaN                                  AA   
3                -0.25        NaN                                  AA   
4                -0.25        NaN                                  AA   
..                 ...        ...                 ...             ...   
213               4.00        NaN                                   P   
232               4.33        NaN                                   Q   
234               4.00        NaN                                   Q   
239               4.67        NaN                                   R   
244               5.67        NaN                                   U   

    Grade Level Equiv.2 Inst Level.1 Grade Level Equiv.3  Accuracy (%)  \
0                 -0.25            A                 0.0          88.0   
1                 -0.25            A                 0.0          97.0   
2                 -0.25            A                 0.0          33.0   
3                                 AA               -0.25           NaN   
4                                 AA               -0.25           NaN   
..                  ...          ...                 ...           ...   
213                 4.0            Q                4.33          99.0   
232                4.33            S                 5.0          99.0   
234                4.33            R                4.67          99.0   
239                4.67            S                 5.0          98.0   
244                5.67            V                 6.0          97.0   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
0              1.0                3.0                2.0        NaN   
1              2.0                2.0                1.0        NaN   
2              2.0                1.0                2.0        NaN   
3              3.0                1.0                NaN        NaN   
4              2.0                3.0                1.0        NaN   
..             ...                ...                ...        ...   
213            3.0                2.0                2.0        1.0   
232            2.0                3.0                2.0        1.0   
234            2.0                3.0                2.0        1.0   
239            2.0                2.0                2.0        2.0   
244            2.0                1.0                1.0        0.0   

     Total Comp  
0           5.0  
1           3.0  
2           3.0  
3           1.0  
4           4.0  
..          ...  
213         5.0  
232         6.0  
234         6.0  
239         6.0  
244         2.0  

[87 rows x 17 columns]

In [15]:
df['Grade Level Equiv'].max()

5.67

The data shows that the instructional level is typically one level above the BOY Indep Level. 5.67 also appears to be highest level, which is equivalent to level U.

In [16]:
# Filling Instructional Level given BOY Independent Levels
i = 0
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['BOY Indep Level']==alpha_keys[i]), 'Inst Level']=alpha_keys[i+1]
        i=+1

In [17]:
# Filling null values:
df['Grade Level Equiv'].fillna('NE', inplace=True)

# Verifying the filling functioned correctly.
df.loc[df['Grade Level Equiv'].isna()]

Empty DataFrame
Columns: [Student Number, Homeroom, Grade Level, BOY Indep Level, Grade Level Equiv, Inst Level, Grade Level Equiv.1, MOY Indep Level, Grade Level Equiv.2, Inst Level.1, Grade Level Equiv.3, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [18]:
# Verifying there are no records with missing BOY Inst Level data
df.loc[df['Inst Level'].isna()]

Student Number    Homeroom  Grade Level BOY Indep Level Grade Level Equiv  \
67     238231914.0  Swarthmore          1.0             NaN                NE   

   Inst Level Grade Level Equiv.1 MOY Indep Level Grade Level Equiv.2  \
67        NaN                                   D                 1.0   

   Inst Level.1 Grade Level Equiv.3  Accuracy (%)  Fluency (0-3)  \
67            E                1.16           NaN            1.0   

    Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
67                1.0                2.0        NaN         3.0

In [19]:
# Filling null values:
df['Grade Level Equiv'].fillna('NE', inplace=True)

# Verifying the filling functioned correctly.
df.loc[df['Grade Level Equiv'].isna()]

df.loc[(df['Grade Level Equiv']=='NE'), 'BOY Indep Level']='NE'
df.loc[(df['Grade Level Equiv']=='NE'), 'Inst Level']='NE'
df.loc[(df['Grade Level Equiv']=='NE'), 'Grade Level Equiv']='NE'
df.loc[(df['Grade Level Equiv']=='NE'), 'Grade Level Equiv.1']='NE'

df.loc[df['Grade Level Equiv']=='NE']

Student Number    Homeroom  Grade Level BOY Indep Level  \
24      240415372.0    Syracuse          0.0              NE   
26      239668965.0    Syracuse          0.0              NE   
53      240905430.0      Howard          0.0              NE   
60      205888340.0  Swarthmore          1.0              NE   
67      238231914.0  Swarthmore          1.0              NE   
73      237499462.0  Swarthmore          1.0              NE   
82      234292928.0  Swarthmore          1.0              NE   
83      218702165.0  Swarthmore          1.0              NE   
138     234743243.0    Williams          2.0              NE   
139     234411106.0    Williams          2.0              NE   
175     226787604.0    Tuskegee          3.0              NE   
189     233199898.0    Tuskegee          3.0              NE   
190     231197146.0    Tuskegee          3.0              NE   
230     232795195.0  Georgetown          4.0              NE   

    Grade Level Equiv Inst Level Grade Level Equiv.1 MOY Indep Level  \
24                 NE         NE                  NE              AA   
26                 NE         NE                  NE              AA   
53                 NE         NE                  NE                   
60                 NE         NE                  NE               C   
67                 NE         NE                  NE               D   
73                 NE         NE                  NE               F   
82                 NE         NE                  NE                   
83                 NE         NE                  NE                   
138                NE         NE                  NE                   
139                NE         NE                  NE                   
175                NE         NE                  NE               M   
189                NE         NE                  NE                   
190                NE         NE                  NE                   
230                NE         NE                  NE               Q   

    Grade Level Equiv.2 Inst Level.1 Grade Level Equiv.3  Accuracy (%)  \
24                                 A                 NaN           NaN   
26                                 A                 NaN           NaN   
53                                                   NaN           NaN   
60                 0.67            D                 1.0          99.0   
67                  1.0            E                1.16           NaN   
73                 1.33            G                 1.5          94.0   
82                                                   NaN           NaN   
83                                                   NaN           NaN   
138                                                  NaN           NaN   
139                                                  NaN           NaN   
175                 3.0            N                 NaN          99.0   
189                                                  NaN           NaN   
190                                                  NaN           NaN   
230                4.33            R                4.67          99.0   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
24             NaN                NaN                NaN              
26             NaN                NaN                NaN              
53             NaN                NaN                NaN              
60             2.0                2.0                2.0        NaN   
67             1.0                1.0                2.0        NaN   
73             1.0                NaN                1.0        NaN   
82             NaN                NaN                NaN              
83             NaN                NaN                NaN              
138            NaN                NaN                NaN              
139            NaN                NaN                NaN              
175            2.0                2.0                2.0        1.0   


In [20]:
# Addressing Grade Level Equiv.1 for Instructional Level
for level, equiv in levels.items():
    df.loc[(df['Inst Level']==level), 'Grade Level Equiv.1'] = equiv
    
# Checking for accuracy    
df.loc[df['Inst Level'].notnull()]

Student Number   Homeroom  Grade Level BOY Indep Level Grade Level Equiv  \
0       236563003.0   Syracuse          0.0              AA             -0.25   
1       238194344.0   Syracuse          0.0              AA             -0.25   
2       239980246.0   Syracuse          0.0              AA             -0.25   
3       240845834.0   Syracuse          0.0              AA             -0.25   
4       239980196.0   Syracuse          0.0              AA             -0.25   
..              ...        ...          ...             ...               ...   
269     233376235.0  Wisconsin          4.0               Q              4.33   
270     232795112.0  Wisconsin          4.0               Q              4.33   
271     224448902.0  Wisconsin          4.0               R              4.67   
272     232836288.0  Wisconsin          4.0               R              4.67   
273     231019795.0  Wisconsin          4.0               R              4.67   

    Inst Level Grade Level Equiv.1 MOY Indep Level Grade Level Equiv.2  \
0            A                 0.0              AA               -0.25   
1            A                 0.0              AA               -0.25   
2            A                 0.0              AA               -0.25   
3            A                 0.0              AA                       
4            A                 0.0              AA                       
..         ...                 ...             ...                 ...   
269          R                4.67               R                4.67   
270          R                4.67               R                4.67   
271          S                 5.0               S                 5.0   
272          S                 5.0               S                 5.0   
273          S                 5.0               T                5.33   

    Inst Level.1 Grade Level Equiv.3  Accuracy (%)  Fluency (0-3)  \
0              A                 0.0          88.0            1.0   
1              A                 0.0          97.0            2.0   
2              A                 0.0          33.0            2.0   
3             AA               -0.25           NaN            3.0   
4             AA               -0.25           NaN            2.0   
..           ...                 ...           ...            ...   
269            S                 5.0          99.0            2.0   
270            S                 5.0          98.0            2.0   
271            T                5.33          98.0            2.0   
272            T                5.33          99.0            3.0   
273            U                5.67          97.0            2.0   

     Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
0                  3.0                2.0        NaN         5.0  
1                  2.0                1.0        NaN         3.0  
2                  1.0                2.0        NaN         3.0  
3                  1.0                NaN        NaN         1.0  
4                  3.0                1.0        NaN         4.0  
..                 ...                ...        ...         ...  
269                2.0                2.0        2.0         6.0  
270                2.0                1.0        3.0         6.0  
271                2.0                2.0        2.0         6.0  
272                2.0                2.0        1.0         5.0  
273                2.0                2.0        2.0         6.0  

[274 rows x 17 columns]

In [21]:
# Locating records with missing MOY Indepent Level data
df.loc[df['MOY Indep Level'].isna()]

Student Number Homeroom  Grade Level BOY Indep Level Grade Level Equiv  \
52      234033942.0   Howard          0.0              AA             -0.25   
110     234623817.0  Spelman          1.0               B              0.33   
111     236417994.0  Spelman          1.0              AA             -0.25   

    Inst Level Grade Level Equiv.1 MOY Indep Level Grade Level Equiv.2  \
52           A                 0.0             NaN                       
110          C                0.67             NaN                       
111          A                 0.0             NaN                       

    Inst Level.1 Grade Level Equiv.3  Accuracy (%)  Fluency (0-3)  \
52             A                 0.0           NaN            1.0   
110            D                 1.0          97.0            1.0   
111            D                 1.0          96.0            1.0   

     Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
52                 3.0                1.0        NaN         4.0  
110                3.0                2.0        NaN         5.0  
111                3.0                3.0        NaN         6.0

The Mid-Year IP Level is missing a few MOY Independent Levels. The MOY Independent Levels will be filled using these instructional levels.

In [22]:
# Filling MOY Independent Levels given Instructional Level
i = 1
for i in range(len(alpha_keys)-1):
    if (df['MOY Indep Level'].isna().any()):
        for level in alpha_keys:
            df.loc[(df['Inst Level.1']==alpha_keys[i]), 'MOY Indep Level']=alpha_keys[i-1]
            i=+1

In [23]:
# Checking for accuracy    
df.loc[df['MOY Indep Level'].isna()]

Empty DataFrame
Columns: [Student Number, Homeroom, Grade Level, BOY Indep Level, Grade Level Equiv, Inst Level, Grade Level Equiv.1, MOY Indep Level, Grade Level Equiv.2, Inst Level.1, Grade Level Equiv.3, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [24]:
df['MOY Indep Level'].shape

(274,)

We now have 274 complete MOY Independent Levels. I will move on to investigate other columns.

In [25]:
df.loc[df['Grade Level Equiv.2'].isna()]

Empty DataFrame
Columns: [Student Number, Homeroom, Grade Level, BOY Indep Level, Grade Level Equiv, Inst Level, Grade Level Equiv.1, MOY Indep Level, Grade Level Equiv.2, Inst Level.1, Grade Level Equiv.3, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [26]:
df.loc[df['Grade Level Equiv.2'].isna()].shape

(0, 17)

In [27]:
# Addressing Grade Level Equiv.2 for Instructional Level
for level, equiv in levels.items():
    df.loc[(df['MOY Indep Level']==level), 'Grade Level Equiv.2'] = equiv
    
# Checking for accuracy    
df.loc[df['Grade Level Equiv.2'].isna()].shape

(0, 17)

In [28]:
# Double checking NE students.

df.loc[df['BOY Indep Level']=="NE"]

Student Number    Homeroom  Grade Level BOY Indep Level  \
24      240415372.0    Syracuse          0.0              NE   
26      239668965.0    Syracuse          0.0              NE   
53      240905430.0      Howard          0.0              NE   
60      205888340.0  Swarthmore          1.0              NE   
67      238231914.0  Swarthmore          1.0              NE   
73      237499462.0  Swarthmore          1.0              NE   
82      234292928.0  Swarthmore          1.0              NE   
83      218702165.0  Swarthmore          1.0              NE   
138     234743243.0    Williams          2.0              NE   
139     234411106.0    Williams          2.0              NE   
175     226787604.0    Tuskegee          3.0              NE   
189     233199898.0    Tuskegee          3.0              NE   
190     231197146.0    Tuskegee          3.0              NE   
230     232795195.0  Georgetown          4.0              NE   

    Grade Level Equiv Inst Level Grade Level Equiv.1 MOY Indep Level  \
24                 NE         NE                  NE              AA   
26                 NE         NE                  NE              AA   
53                 NE         NE                  NE                   
60                 NE         NE                  NE               C   
67                 NE         NE                  NE               D   
73                 NE         NE                  NE               F   
82                 NE         NE                  NE                   
83                 NE         NE                  NE                   
138                NE         NE                  NE                   
139                NE         NE                  NE                   
175                NE         NE                  NE               M   
189                NE         NE                  NE                   
190                NE         NE                  NE                   
230                NE         NE                  NE               Q   

    Grade Level Equiv.2 Inst Level.1 Grade Level Equiv.3  Accuracy (%)  \
24                -0.25            A                 NaN           NaN   
26                -0.25            A                 NaN           NaN   
53                                                   NaN           NaN   
60                 0.67            D                 1.0          99.0   
67                  1.0            E                1.16           NaN   
73                 1.33            G                 1.5          94.0   
82                                                   NaN           NaN   
83                                                   NaN           NaN   
138                                                  NaN           NaN   
139                                                  NaN           NaN   
175                 3.0            N                 NaN          99.0   
189                                                  NaN           NaN   
190                                                  NaN           NaN   
230                4.33            R                4.67          99.0   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
24             NaN                NaN                NaN              
26             NaN                NaN                NaN              
53             NaN                NaN                NaN              
60             2.0                2.0                2.0        NaN   
67             1.0                1.0                2.0        NaN   
73             1.0                NaN                1.0        NaN   
82             NaN                NaN                NaN              
83             NaN                NaN                NaN              
138            NaN                NaN                NaN              
139            NaN                NaN                NaN              
175            2.0                2.0                2.0        1.0   


In [29]:
df.dropna(subset=['Grade Level Equiv.2'], inplace=True)

# Checking for accuracy
df.shape

(274, 17)

In [30]:
# Checking for accuracy    
df.loc[df['Grade Level Equiv.2'].isna()].shape

(0, 17)

In [31]:
df.shape

(274, 17)

In [32]:
# Next column: Instructional Level.
df.loc[df['Inst Level.1'].isna()].shape

(10, 17)

In [33]:
# Investigating 10 missing records.
df.loc[df['Inst Level.1'].isna()]

Student Number    Homeroom  Grade Level BOY Indep Level  \
18      239980154.0    Syracuse          0.0              AA   
54      237299441.0  Swarthmore          1.0               A   
153     234623676.0    Brooklyn          2.0               G   
221     228040739.0  Georgetown          4.0               J   
222     223297714.0  Georgetown          4.0               J   
228     209155860.0  Georgetown          4.0               M   
229     223414517.0  Georgetown          4.0               N   
235     231770470.0  Georgetown          4.0               Q   
236     230879421.0  Georgetown          4.0               P   
242     223297771.0  Georgetown          4.0               R   

    Grade Level Equiv Inst Level Grade Level Equiv.1 MOY Indep Level  \
18              -0.25          A                 0.0               C   
54                0.0          B                0.33              AA   
153               1.5          H                1.67               J   
221               2.0          K                2.33               L   
222               2.0          K                2.33               M   
228               3.0          N                3.33               P   
229              3.33          O                3.67               P   
235              4.33          R                4.67               Q   
236               4.0          Q                4.33               Q   
242              4.67          S                 5.0               S   

    Grade Level Equiv.2 Inst Level.1 Grade Level Equiv.3  Accuracy (%)  \
18                 0.67          NaN                 NaN           NaN   
54                -0.25          NaN                 NaN           NaN   
153                 2.0          NaN                 NaN           NaN   
221                2.67          NaN                               NaN   
222                 3.0          NaN                               NaN   
228                 4.0          NaN                               NaN   
229                 4.0          NaN                 NaN           NaN   
235                4.33          NaN                               NaN   
236                4.33          NaN                               NaN   
242                 5.0          NaN                               NaN   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
18             NaN                NaN                NaN        NaN   
54             NaN                NaN                NaN        NaN   
153            NaN                NaN                NaN        NaN   
221            NaN                NaN                NaN        NaN   
222            NaN                NaN                NaN        NaN   
228            NaN                NaN                NaN        NaN   
229            NaN                NaN                NaN        NaN   
235            NaN                NaN                NaN        NaN   
236            NaN                NaN                NaN        NaN   
242            NaN                NaN                NaN        NaN   

     Total Comp  
18          0.0  
54          0.0  
153         0.0  
221         0.0  
222         0.0  
228         0.0  
229         0.0  
235         0.0  
236         0.0  
242         0.0

In [34]:
# Filling MOY Instructional Level given MOY Independent Levels
i = 0
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['MOY Indep Level']==alpha_keys[i]), 'Inst Level.1']=alpha_keys[i+1]
        i=+1

# Checking for completeness
df.loc[df['Inst Level.1'].isna()]

Empty DataFrame
Columns: [Student Number, Homeroom, Grade Level, BOY Indep Level, Grade Level Equiv, Inst Level, Grade Level Equiv.1, MOY Indep Level, Grade Level Equiv.2, Inst Level.1, Grade Level Equiv.3, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [35]:
# Filling Grade Level Equiv.3 given Instructional Levels
i = 0
for level, equiv in levels.items():
    df.loc[(df['Inst Level.1']==level), 'Grade Level Equiv.3'] = equiv
    
# Checking for completeness
df.loc[df['Grade Level Equiv.3'].isna()]

Student Number    Homeroom  Grade Level BOY Indep Level  \
53      240905430.0      Howard          0.0              NE   
82      234292928.0  Swarthmore          1.0              NE   
83      218702165.0  Swarthmore          1.0              NE   
138     234743243.0    Williams          2.0              NE   
139     234411106.0    Williams          2.0              NE   
189     233199898.0    Tuskegee          3.0              NE   
190     231197146.0    Tuskegee          3.0              NE   

    Grade Level Equiv Inst Level Grade Level Equiv.1 MOY Indep Level  \
53                 NE         NE                  NE                   
82                 NE         NE                  NE                   
83                 NE         NE                  NE                   
138                NE         NE                  NE                   
139                NE         NE                  NE                   
189                NE         NE                  NE                   
190                NE         NE                  NE                   

    Grade Level Equiv.2 Inst Level.1 Grade Level Equiv.3  Accuracy (%)  \
53                                                   NaN           NaN   
82                                                   NaN           NaN   
83                                                   NaN           NaN   
138                                                  NaN           NaN   
139                                                  NaN           NaN   
189                                                  NaN           NaN   
190                                                  NaN           NaN   

     Fluency (0-3)  Comp Within (0-3)  Comp Beyond (0-3) Comp About  \
53             NaN                NaN                NaN              
82             NaN                NaN                NaN              
83             NaN                NaN                NaN              
138            NaN                NaN                NaN              
139            NaN                NaN                NaN              
189            NaN                NaN                NaN              
190            NaN                NaN                NaN              

     Total Comp  
53          0.0  
82          0.0  
83          0.0  
138         0.0  
139         0.0  
189         0.0  
190         0.0

The data shows that students missing MOY data also were students who are also missing in the BOY. My assumption is that these students were registered, but didnot attend the school. I will drop these records from the data frame.

In [41]:
df.dropna(subset=['Grade Level Equiv.3'], inplace=True)

In [43]:
df['Grade Level Equiv.3'].isna().sum()

0

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 267 entries, 0 to 273
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Student Number       267 non-null    float64
 1   Homeroom             267 non-null    object 
 2   Grade Level          267 non-null    float64
 3   BOY Indep Level      267 non-null    object 
 4   Grade Level Equiv    260 non-null    float64
 5   Inst Level           267 non-null    object 
 6   Grade Level Equiv.1  260 non-null    float64
 7   MOY Indep Level      267 non-null    object 
 8   Grade Level Equiv.2  267 non-null    float64
 9   Inst Level.1         267 non-null    object 
 10  Grade Level Equiv.3  267 non-null    float64
 11  Accuracy (%)         218 non-null    float64
 12  Fluency (0-3)        232 non-null    float64
 13  Comp Within (0-3)    251 non-null    float64
 14  Comp Beyond (0-3)    249 non-null    float64
 15  Comp About           125 non-null    obj

In [37]:
# Removing students NE and blank MOY.


This first grader's BOY Independent Level was A and MOY Independent Level was Z. Such a high jump is highly unlikely. An investigation into these outliers are required.

In [38]:
# Changing data to numerical type
df[['Grade Level Equiv','Grade Level Equiv.1','Grade Level Equiv.2','Grade Level Equiv.3']] = df[['Grade Level Equiv','Grade Level Equiv.1','Grade Level Equiv.2','Grade Level Equiv.3']].apply(pd.to_numeric, errors='coerce')

In [39]:
# Investigating Outliers by creating a column showing the difference between BOY and MOY levels.
df.isna().sum()
df['BOY-MOY Growth'] = abs(df['Grade Level Equiv.2'] - df['Grade Level Equiv'])
df['BOY-MOY Growth']

0      0.00
1      0.00
2      0.00
3      8.25
4      8.25
       ... 
269    0.34
270    0.34
271    0.33
272    0.33
273    0.66
Name: BOY-MOY Growth, Length: 274, dtype: float64

Since this analysis compares student's reading levels from the beginning of the school to the middle of school year, I expect to see a few cases of regression, no growth, and atleast a half a year of growth in reading levels. According to F&P Instructional Level Expectations for Reading, students can grow 3-4 levels within a school year. For example, a first grader is expected to enter the grade reading on level E or above and be on level K or above by the end of the school year. For this part of the analysis, I will isolate students who grew more than two grade levels and investigate their growth.

In [40]:
for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY Indep Level']}, {row['MOY Indep Level']}")
    

Index: 0, AA, AA
Index: 1, AA, AA
Index: 2, AA, AA
Index: 3, AA, Z
Index: 4, AA, Z
Index: 5, AA, A
Index: 6, AA, A
Index: 7, AA, A
Index: 8, AA, A
Index: 9, AA, A
Index: 10, AA, A
Index: 11, AA, B
Index: 12, AA, B
Index: 13, AA, B
Index: 14, AA, B
Index: 15, AA, D
Index: 16, AA, B
Index: 17, AA, C
Index: 18, AA, C
Index: 19, AA, C
Index: 20, AA, D
Index: 21, B, D
Index: 22, AA, D
Index: 23, AA, F
Index: 24, NE, AA
Index: 25, AA, B
Index: 26, NE, AA
Index: 27, AA, AA
Index: 28, AA, Z
Index: 29, AA, Z
Index: 30, AA, Z
Index: 31, AA, Z
Index: 32, AA, A
Index: 33, AA, A
Index: 34, AA, A
Index: 35, AA, A
Index: 36, AA, A
Index: 37, AA, A
Index: 38, AA, A
Index: 39, AA, A
Index: 40, AA, B
Index: 41, AA, B
Index: 42, AA, B
Index: 43, AA, C
Index: 44, AA, C
Index: 45, AA, C
Index: 46, AA, C
Index: 47, AA, C
Index: 48, AA, D
Index: 49, AA, D
Index: 50, AA, H
Index: 51, AA, H
Index: 52, AA, AA
Index: 53, NE,  
Index: 54, A, AA
Index: 55, AA, AA
Index: 56, B, B
Index: 57, B, C
Index: 58, AA, A
In